File: examples/03-retry-catch/handlers/handlers.go
package handlers

import (
	"context"
	"fmt"

	"github.com/pastorenue/kflow/pkg/kflow"
)

// PipelineHandlers simulates a flaky external service.
// succeedAfter == 0 means always fail (triggers catch after retries exhausted).
type PipelineHandlers struct {
	succeedAfter int
	attempt      int
}

func New(succeedAfter int) *PipelineHandlers {
	return &PipelineHandlers{succeedAfter: succeedAfter}
}

func (h *PipelineHandlers) FetchExternalData(_ context.Context, input kflow.Input) (kflow.Output, error) {
	h.attempt++
	fmt.Printf("  FetchExternalData: attempt %d\n", h.attempt)
	if h.succeedAfter == 0 || h.attempt < h.succeedAfter {
		return nil, fmt.Errorf("upstream timeout (attempt %d)", h.attempt)
	}
	fmt.Println("  FetchExternalData: success")
	return kflow.Output{
		"record_id": input["record_id"],
		"raw":       map[string]any{"value": 42, "source": "api"},
	}, nil
}

func (h *PipelineHandlers) EnrichRecord(_ context.Context, input kflow.Input) (kflow.Output, error) {
	raw, _ := input["raw"].(map[string]any)
	fmt.Printf("  EnrichRecord: enriching record_id=%v\n", input["record_id"])
	return kflow.Output{
		"record_id": input["record_id"],
		"enriched":  map[string]any{"value": raw["value"], "label": "verified", "source": raw["source"]},
	}, nil
}

func (h *PipelineHandlers) Persist(_ context.Context, input kflow.Input) (kflow.Output, error) {
	fmt.Printf("  Persist: saved record_id=%v\n", input["record_id"])
	return kflow.Output{"record_id": input["record_id"], "persisted": true}, nil
}

func (h *PipelineHandlers) HandleFetchError(_ context.Context, input kflow.Input) (kflow.Output, error) {
	fmt.Printf("  HandleFetchError: caught error=%v\n", input["_error"])
	return kflow.Output{"record_id": input["record_id"], "status": "fetch_failed"}, nil
}

File: examples/03-retry-catch/python/handlers.py
STATE_FETCH_EXTERNAL_DATA = "FetchExternalData"
STATE_ENRICH_RECORD       = "EnrichRecord"
STATE_PERSIST             = "Persist"
STATE_HANDLE_FETCH_ERROR  = "HandleFetchError"


class PipelineHandlers:
    """Simulates a flaky external service. succeed_after=0 means always fail."""

    def __init__(self, succeed_after: int):
        self._succeed_after = succeed_after
        self._attempt = 0

    def fetch_external_data(self, input: dict) -> dict:
        self._attempt += 1
        print(f"  FetchExternalData: attempt {self._attempt}")
        if self._succeed_after == 0 or self._attempt < self._succeed_after:
            raise RuntimeError(f"upstream timeout (attempt {self._attempt})")
        print("  FetchExternalData: success")
        return {
            "record_id": input["record_id"],
            "raw":       {"value": 42, "source": "api"},
        }

    def enrich_record(self, input: dict) -> dict:
        raw = input["raw"]
        print(f"  EnrichRecord: enriching record_id={input['record_id']}")
        return {
            "record_id": input["record_id"],
            "enriched":  {"value": raw["value"], "label": "verified", "source": raw["source"]},
        }

    def persist(self, input: dict) -> dict:
        print(f"  Persist: saved record_id={input['record_id']}")
        return {"record_id": input["record_id"], "persisted": True}

    def handle_fetch_error(self, input: dict) -> dict:
        print(f"  HandleFetchError: caught error={input.get('_error')}")
        return {"record_id": input["record_id"], "status": "fetch_failed"}
